### PR TITLE
onMouseOver and onMouseLeave options, resolves #138

### LINKED ISSUE #138: On hover
One of the next enhancements I'm interested in is the ability to add callbacks for hovering over properties and values. Do you think this addition would be suitable for the core library?

I'm thinking something relatively straightforward for the interface:

- onPropMouseOver: callback for hover into JSON object properties
- onPropMouseOut: callback for hover out JSON object propperties
- onValueMouseOver: callback for hovering in over JSON object values
- onValueMouseOut: callback for hover out of JSON object value

However if you have a preference or better idea I'm happy to hear it.

### PR DESCRIPTION
**Pull request by ajbdev**
I realized for my use case I don't need separate callbacks for the property and value hovers, so I figured I'd implement just onMouseEnter and onMouseLeave so it's simpler and a little more uniform to the other callback function options.

Let me know what you think.

### CODE REVIEW

**Comment by mac-s-g** [comment]
nice, i'll check it out as soon as I can!

**Comment by mac-s-g** [comment]
looks good! and thank you for including tests.

two things:

1. merge the master branch in.  you'll need to re-install dependencies once that's done because they were updated recently.  I verified that this will get the build passing.
2.  can you abstract the `onMouseEnter` and `onMouseLeave` handler definitions into a single component?

I'd recommend creating a new file called `VariableRow.js`, and creating a component like this:
```
import React from "react";

export default ({
  onMouseEnter,
  onMouseLeave,
  namespace,
  name,
  src,
  children,
  ...rest
}) => (
  <div
    {...rest}
    onMouseEnter={
      !onMouseEnter
        ? null
        : () => {
            const variable = new JsonVariable(name, src);
            let location = [...namespace];
            location.shift();
            location.pop();
            onMouseEnter({
              ...variable,
              namespace: location
            });
          }
    }
    onMouseLeave={
      !onMouseLeave
        ? null
        : () => {
            const variable = new JsonVariable(name, src);
            let location = [...namespace];
            location.shift();
            location.pop();
            onMouseEnter({
              ...variable,
              namespace: location
            });
          }
    }
  >
    {children}
  </div>
);
```

that way we don't have to worry about maintaining those handlers in two separate areas.  

let me know if you have any issues with that and I can post a pr with more detail.

overall, the code looks awesome!  thanks again for contributing.

File: src/js/components/DataTypes/Object.js [hunk 1]
```diff
@@ -174,10 +175,42 @@ class rjvObject extends React.Component {
...
+                }
+                onMouseLeave={
+                    !onMouseLeave
+                        ? null
+                        : () => { 
+                            let location = [...namespace]
+                            location.shift()
+                            location.pop()
+                            const variable = new JsonVariable(name, src)
+                            onMouseEnter({
```
Comment (mac-s-g) [inline src/js/components/DataTypes/Object.js]: this should call `onMouseLeave`

**Comment by mac-s-g** [comment]
@ajbdev 

I migrated your changes to a branch here:
https://github.com/mac-s-g/react-json-view/compare/master...variable-hover-callbacks

I made the changes I requested above and verified that tests are passing.

If you want, you can merge my branch into this branch and update the PR.  I can also just create a PR with that branch - but I'd like you to get credit for the contributions.

**Comment by ajbdev** [comment]
Awesome! Thank you! I'll fetch your branch and add it to this PR.

**Comment by mac-s-g** [comment]
I've been keeping #158 up to date so I'm gonna close this one.

**Pull request closed without merging**